DOC: mention generalized ufuncs, document signature attribute

mhvk opened the pull request:
While trying to add documentation for the new `axes` argument for generalized ufuncs (#8819), I realized that the current documentation does not actually discuss generalized ufuncs at all (except in the C API section), and that the signature attribute is not documented. This PR attempts to correct that. Since we do not expose any generalized ufuncs yet (though may soon; #8528), I kept it brief.

As a bonus, now the ufunc section also mentions `__array_ufunc__`...

p.s. I seem unable to generate documentation (segmentation faults), so am not able to test the links...

mhvk committed:
DOC: Update ufunc documentation to mention generalized ufuncs.

Also ensure that __array_ufunc__ is referenced in the output type
determination.

mhvk committed:
DOC: add docstring to ufunc.signature

charris commented on doc/source/reference/ufuncs.rst:
```diff
@@ -322,7 +333,8 @@ advanced usage and will not typically be used.
     Values of True indicate to calculate the ufunc at that position, values
-    of False indicate to leave the value in the output alone.
+    of False indicate to leave the value in the output alone. This argument
+    cannot be used for generalized ufuncs (as these take non-scalar input).
```
Don't need either parenthesis or comma here when "as" is used with the meaning of "because".

charris replied:
Don't need <- should omit.

eric-wieser commented on numpy/add_newdocs.py:
```diff
@@ -5663,6 +5663,36 @@ def luf(lamdaexpr, *args, **kwargs):
+    Generalized ufuncs are used internally in many linalg functions, and in
+    the testing suite; the examples below are taken from these.
+    For ufuncs that operate on scalars, the signature is `None`, which is
+    equivalent to '()' for every argument.
```
Probably worth noting somewhere that `ufunc.__call__(..., signature='dd->d')` is a completely different meaning of signature to the `.signature` property

charris replied:
I added a note to the `signature` entry.

charris committed:
MAINT: Fix use of parenthesis. 

[ci skip]

charris committed:
DOC:: Clarify the two ufunc signature types.

The `ufunc.signature` attribute is the generalized ufunc signature, while the
optional `signature` parameter in `ufunc.__call__` refers to the loop type.

[ci skip].

charris commented:
Thanks Marten.

mhvk commented:
Oops, was just about to push another bit of text in the docstring for the attribute. But probably your text is good enough! Thanks!

charris commented:
@mhvk Sorry for stepping on your toes. I want to branch 1.14 today, so getting a bit "pushy" :)
